scripts/knot_benchmark/src/benchmark/cases.py

```diff
@@ -0,0 +1,223 @@
+            case _:
+                raise ValueError(f"Invalid benchmark: {benchmark}")
+
+    @abc.abstractmethod
+    def cold_command(self, project: Project, venv: Venv) -> Command: ...
+
+    def warm_command(self, project: Project, venv: Venv) -> Command | None:
+        return None
+
+    def incremental_command(self, project: Project, venv: Venv) -> Command | None:
+        return None
+
+
+class Knot(Tool):
+    path: Path
+    name: str
+
+    def __init__(self, *, path: Path | None = None):
+        self.name = str(path) or "knot"
+        self.path = path or (
+            Path(__file__)
+            .joinpath(
+                "../../../../../",
+                "target",
+                "release",
+                "red_knot",
+            )
+            .resolve()
+        )
+
+        assert self.path.is_file(), f"Red Knot not found at '{self.path}'. Run `cargo build --release --bin red_knot`."
+
+    def cold_command(self, project: Project, venv: Venv) -> Command:
+        command = [str(self.path), "-v"]
+
+        assert len(project.include) < 2, "Knot doesn't support multiple source folders"
+
+        if directory := project.include[0]:
+            command.extend(["--current-directory", directory])
+
+        command.extend(["--venv-path", str(venv.path)])
+
+        return Command(
+            name="knot",
+            command=command,
+        )
+
+
+class MyPy(Tool):
+    path: Path
+
+    def __init__(self, *, path: Path | None = None):
+        self.path = path or which_tool(
+            "mypy",
+        )
+
+    def cold_command(self, project: Project, venv: Venv) -> Command:
+        command = [
+            *self._base_command(project, venv),
+            "--no-incremental",
+            "--cache-dir",
+            os.devnull,
+        ]
+
+        return Command(
+            name="mypy",
+            command=command,
+        )
+
+    def warm_command(self, project: Project, venv: Venv) -> Command | None:
+        command = [
+            str(self.path),
+            *(project.mypy_arguments or project.include),
+            "--python-executable",
+            str(venv.python),
+        ]
+
+        return Command(
+            name="mypy",
+            command=command,
+        )
+
+    def _base_command(self, project: Project, venv: Venv) -> list[str]:
+        return [
+            str(self.path),
+            "--python-executable",
+            str(venv.python),
+            *(project.mypy_arguments or project.include),
+        ]
+
+
+class Pyright(Tool):
+    path: Path
+
+    def __init__(self, *, path: Path | None = None):
+        self.path = path or which_tool("pyright")
+
+    def cold_command(self, project: Project, venv: Venv) -> Command:
+        command = [
+            str(self.path),
+            "--venvpath",
+            str(
+                venv.path.parent
+            ),  # This is not the path to the venv folder, but the folder that contains the venv...
```

Comment:
wow, bizarre. Do we also need to specify `--venv=venv` along with this argument, so that it knows which subdirectory inside the `--venv-path` directory the venv is in? https://github.com/microsoft/pyright/blob/main/docs/import-resolution.md#configuring-your-python-environment

Reply:
Yes we do. But there's no `venv` CLI option. So we have to create a config here 

https://github.com/astral-sh/ruff/blob/28debb8d0667065940b9a2d288b164f3efd06a2d/src/benchmark/run.py#L126-L129

Reply:
pyright's config setup is so odd...